File: Design Patterns/Factory/factory.py
from abc import ABC, abstractmethod

###########################################
# Product Interface: Shape
###########################################

class Shape(ABC):
    @abstractmethod
    def draw(self):
        pass

###########################################
# Concrete Products: Shapes
###########################################

class Rectangle(Shape):
    def draw(self):
        print("Rectangle")

class Square(Shape):
    def draw(self):
        print("Square")

class Circle(Shape):
    def draw(self):
        print("Circle")

###########################################
# Factory Class: ShapeFactory
###########################################

class ShapeFactory:
    def get_shape(self, shape_type: str) -> Shape | None:
        shape_type = shape_type.lower()
        if shape_type == "rectangle":
            return Rectangle()
        elif shape_type == "square":
            return Square()
        elif shape_type == "circle":
            return Circle()
        else:
            return None

###########################################
# Main Function: Demonstration
###########################################

if __name__ == "__main__":
    print("Factory Pattern")

    shape_factory = ShapeFactory()

    shape1 = shape_factory.get_shape("Circle")
    shape1.draw()  # Output: Circle

    shape2 = shape_factory.get_shape("Rectangle")
    shape2.draw()  # Output: Rectangle

    shape3 = shape_factory.get_shape("Square")
    shape3.draw()  # Output: Square

File: Design Patterns/Creational/Factory/factory.cpp
#include <bits/stdc++.h>
using namespace std;

///////////////////////////////////////////
// Product Interface: Shape
///////////////////////////////////////////

/*
 * Shape:
 * Abstract base class representing the product.
 * Defines the interface for all concrete shape objects.
 */
class Shape
{
public:
  virtual void draw() = 0; // Pure virtual method
  virtual ~Shape() = default;
};

///////////////////////////////////////////
// Concrete Products: Shapes
///////////////////////////////////////////

/*
 * Rectangle:
 * Concrete implementation of Shape.
 */
class Rectangle : public Shape
{
public:
  void draw() override
  {
    cout << "Rectangle" << endl;
  }
};

/*
 * Square:
 * Concrete implementation of Shape.
 */
class Square : public Shape
{
public:
  void draw() override
  {
    cout << "Square" << endl;
  }
};

/*
 * Circle:
 * Concrete implementation of Shape.
 */
class Circle : public Shape
{
public:
  void draw() override
  {
    cout << "Circle" << endl;
  }
};

///////////////////////////////////////////
// Factory Class: ShapeFactory
///////////////////////////////////////////

/*
 * ShapeFactory:
 * Responsible for creating instances of concrete shapes.
 * The client requests an object by type name (string).
 */
class ShapeFactory
{
public:
  Shape *getShape(const string &shapeType)
  {
    if (shapeType == "Rectangle")
    {
      return new Rectangle();
    }
    else if (shapeType == "Square")
    {
      return new Square();
    }
    else if (shapeType == "Circle")
    {
      return new Circle();
    }
    else
    {
      return nullptr;
    }
  }
};

///////////////////////////////////////////
// Main Function: Demonstration
///////////////////////////////////////////

/*
 * Demonstrates the Factory Design Pattern.
 * The factory creates different shape objects based on input.
 */
int main()
{
  cout << "Factory Pattern" << endl;

  // Create factory instance
  ShapeFactory *shapeFactory = new ShapeFactory();

  // Request different shapes from the factory
  Shape *shape1 = shapeFactory->getShape("Circle");
  shape1->draw(); // Output: Circle

  Shape *shape2 = shapeFactory->getShape("Rectangle");
  shape2->draw(); // Output: Rectangle

  Shape *shape3 = shapeFactory->getShape("Square");
  shape3->draw(); // Output: Square

  // Cleanup
  delete shape1;
  delete shape2;
  delete shape3;
  delete shapeFactory;

  return 0;
}
